src/expr/src/relation/func.rs

```diff
@@ -978,6 +978,59 @@ fn generate_series_ts(
     Ok(tsri.map(move |i| (Row::pack_slice(&[conv(i)]), 1)))
 }
 
+fn generate_subscripts_array(
+    a: Datum,
+    dim: i32,
+) -> Result<Box<dyn Iterator<Item = (Row, Diff)>>, EvalError> {
+    if dim <= 0 {
+        return Ok(Box::new(iter::empty()));
+    }
+
+    match a.unwrap_array().dims().into_iter().nth(
+        (dim - 1)
+            .try_into()
+            .expect("array dimensions must be a usize"),
+    ) {
+        Some(requested_dim) => generate_series::<i32>(
+            1,
+            requested_dim
+                .length
+                .try_into()
+                .expect("array dimensions must be a usize"),
+            1,
+        )
+        .map(|e| Box::new(e) as Box<dyn Iterator<Item = _>>),
+        None => Ok(Box::new(iter::empty())),
+    }
+}
+
+fn generate_subscripts_list(
+    a: Datum,
+    dim: i32,
+) -> Result<Box<dyn Iterator<Item = (Row, Diff)>>, EvalError> {
+    a.unwrap_list().iter().try_for_each(|e| match e {
+        Datum::Array(_) | Datum::List(_) => Err(EvalError::InvalidParameterValue(
+            "parsing multi-dimensional lists is not supported".to_owned(),
```

Comment:
This doesn't seem like the right error message, and also I don't think we should error in this case! I left a comment elsewhere, but I think the correct behavior is for `generate_subscripts` on a list to only ever operate on the outermost list, and not take a `dim` arg at all. The correspondence between arrays and lists is pretty forced with fucntions like this.